### PR TITLE
Master new snippet table of content bvr

### PR DESCRIPTION
**Pull request by bvr-odoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            });
+        });
+        this.observer.observe(targetNode, config);
+    },
+    /**
+     * @override
+     */
+    _computeWidgetState: function (methodName, params) {
+        let $headings = '';
+        let $navbar = '';
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/options.js]: Why defining this here, define that where needed and wrap the case clauses in an instructions block.

```javascript
case XXX: {
    const ...
    break;
}
```

File: addons/website/views/snippets/s_table_of_content.xml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+<template id="s_table_of_content" name="Table of Content">
+    <section class="s_table_of_content pt24 pb24">
+        <div class="container table_of_content">
+            <div class="row s_col_no_resize">
+                <div class="col-lg-3 s_table_of_content_navbar_wrap s_table_of_content_vertical_navbar d-print-none d-none d-lg-block o_not_editable bg-white" data-name="Navbar">
+                    <div class="s_table_of_content_navbar list-group"/>
+                </div>
+                <div class="col-lg-9 s_table_of_content_main" data-name="Content">
+                    <div class="oe_structure">
+                        <div class="pb16 pl-lg-4">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/s_table_of_content.xml]: Why is this needed ?

File: addons/website/views/snippets/s_table_of_content.xml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+<template id="s_table_of_content" name="Table of Content">
+    <section class="s_table_of_content pt24 pb24">
+        <div class="container table_of_content">
+            <div class="row s_col_no_resize">
+                <div class="col-lg-3 s_table_of_content_navbar_wrap s_table_of_content_vertical_navbar d-print-none d-none d-lg-block o_not_editable bg-white" data-name="Navbar">
+                    <div class="s_table_of_content_navbar list-group"/>
+                </div>
+                <div class="col-lg-9 s_table_of_content_main" data-name="Content">
+                    <div class="oe_structure">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/s_table_of_content.xml]: Not `oe_empty` ? I don't know

File: addons/website/static/src/snippets/s_table_of_content/000.scss [hunk 4]
```diff
@@ -0,0 +1,55 @@
+.s_table_of_content {
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/000.scss]: Add the `:not([data-vcss])` already

File: addons/website/static/src/snippets/s_table_of_content/000.scss [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    .s_table_of_content_navbar_wrap {
+        &.s_table_of_content_horizontal_navbar, &.s_table_of_content_vertical_navbar .s_table_of_content_navbar {
+            @include o-position-sticky($top: 0px);
+        }
+        &.s_table_of_content_vertical_navbar .s_table_of_content_navbar {
+            & > a.list-group-item-action {
+                background: none;
+                color: rgba($body-color, 0.7);
```
[the remark this answers is not in the document]
Comment (bvr-odoo) [reply addons/website/static/src/snippets/s_table_of_content/000.scss]: by default background is colored and text is white

File: addons/website/views/snippets/s_table_of_content.xml (hunk 2 above)
Comment (bvr-odoo) [reply addons/website/views/snippets/s_table_of_content.xml]: It is not necessary anymore

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    _observeContentChange: function (ev) {
+        const self = this;
+        const targetNode = this.$target.find('.s_table_of_content_main')[0];
+        const config = {attributes: false, childList: true, subtree: true, characterData: true};
+        this.observer = new MutationObserver(function (mutationsList) {
+            _.each(mutationsList, el => {
+                self._generateNav();
+            });
+        });
+        this.observer.observe(targetNode, config);
```
[the remark this answers is not in the document]
Comment (bvr-odoo) [reply addons/website/static/src/snippets/s_table_of_content/options.js]: As discussed, we will keep this observer until we have something better cause the content_changed event does not have exactly the desired behevior in this case

File: addons/website/static/src/js/content/snippets.animation.js [hunk 7]
```diff
@@ -949,7 +953,7 @@ registry.ScrollButton = registry.anchorSlide.extend({
         ev.preventDefault();
         const $nextSection = this.$el.closest('section').next('section');
         if ($nextSection.length) {
-            this._scrollTo($nextSection);
+            this._scrollTo($nextSection, 'true');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/snippets.animation.js]: Could you change the method so that the second parameter is optional and 'true' by default so that there is no need to change here?

File: addons/website/views/website_navbar_templates.xml [hunk 8]
```diff
@@ -6,7 +6,7 @@
             <t t-set="body_classname" t-value="(body_classname if body_classname else '') + (' o_connected_user' if env['ir.ui.view'].user_has_groups('base.group_user') else '')"/>
         </xpath>
         <xpath expr="//div[@id='wrapwrap']" position="before">
-            <nav groups="base.group_user" t-if="website and menu_data" id="oe_main_menu_navbar" class="o_main_navbar">
+            <nav groups="base.group_user" t-if="website and menu_data" id="oe_main_menu_navbar" class="o_main_navbar o_navbar_fixed">
```
Comment (qsm-odoo) [inline addons/website/views/website_navbar_templates.xml]: Rename the class in `o_top_fixed_element`, seems more generic and without associated style.

File: addons/website/static/src/js/content/snippets.animation.js [hunk 9]
```diff
@@ -913,6 +913,15 @@ registry.anchorSlide = publicWidget.Widget.extend({
...
         _.each($navbarFixed, el => headerHeight += $(el).outerHeight());
+        // look for a sticky horizontal navbar if the link comes from a table of content snippet
+        // and add the height of this navbar if it exists
+        if ($el[0].id.startsWith('table_of_content_heading_')) {
+            const $link = $.find('a[href="#' + $el[0].id + '"]');
+            const tableOfContentNavbarEl = $link[0].closest('.s_table_of_content_navbar_sticky.s_table_of_content_horizontal_navbar');
+            if (tableOfContentNavbarEl !== null) {
+                headerHeight += $(tableOfContentNavbarEl).outerHeight();
+            }
+        }
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/snippets.animation.js]: The point of the `o_navbar_fixed` class is to have no specific code here, can't you just add the o_navbar_fixed class on the fixed element of the table of content snippet ? inside the table of content JS ?

File: addons/website/static/src/snippets/s_table_of_content/000.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    },
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * Initialize the top position of the snippet navbar according to the height
+     * of the headers of the page.
+     * @private
```
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_table_of_content/000.js]: @bvr-odoo This is still true, check the remaining comments before sending to review

File: addons/website/static/src/snippets/s_table_of_content/000.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    // Private
+    //--------------------------------------------------------------------------
+    /**
+     * Initialize the top position of the snippet navbar according to the height
+     * of the headers of the page.
+     * @private
+     */
+    _initializeNavbarTopPosition: function () {
+        // TODO: find a way to adjust the horizontal navbar position according
+        // the header height/position when scrolling. (same issue whith s_scrollTo in snippets.animation) Event ?
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/000.js]: I am not sure to understand the problem reading the comment

File: addons/website/static/src/snippets/s_table_of_content/000.scss [hunk 12]
```diff
@@ -0,0 +1,64 @@
+.s_table_of_content:not([data-vcss]) {
+    .s_table_of_content_navbar_wrap {
+        &.s_table_of_content_navbar_sticky {
+            &.s_table_of_content_horizontal_navbar, &.s_table_of_content_vertical_navbar .s_table_of_content_navbar {
+                @include o-position-sticky($top: 0px);
+            }
+        }
+        &:not(.s_table_of_content_navbar_sticky) {
+            &, .s_table_of_content_navbar {
+                top: 0px !important;
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/000.scss]: Why important ?

File: addons/website/static/src/snippets/s_table_of_content/000.scss [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+                    }
+                }
+            }
+        }
+        &.s_table_of_content_horizontal_navbar {
+            z-index: 1;
+            padding-top: 5px;
+            padding-bottom: 5px;
+            margin-bottom: 20px;
+
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/000.scss]: Can't this use bootstrap variables ?

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+     */
+    _observeContentChange: function (ev) {
+        const self = this;
+        const targetNode = this.$target.find('.s_table_of_content_main')[0];
+        const config = {attributes: false, childList: true, subtree: true, characterData: true};
+        this.observer = new MutationObserver(function (mutationsList) {
+            self._generateNav();
+        });
+        this.observer.observe(targetNode, config);
+    },
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/options.js]: No need of a function for this, put this directly in the start and add a comment on top to explain

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+                } else {
+                    return 'true';
+                }
+            }
+            case 'navbarPosition': {
+                const $navbar = this.$target.find('.s_table_of_content_navbar_wrap');
+                if ($navbar.hasClass('s_table_of_content_horizontal_navbar')) {
+                    return 'top';
+                } else {
+                    let $mainContent = this.$target.find('.s_table_of_content_main');
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/options.js]: `const`

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+    /**
+     * @override
+     */
+    start: function () {
+        const leftPanelEl = this.$overlay.data('$optionsSection')[0];
+        leftPanelEl.querySelector('.oe_snippet_remove').classList.add('d-none'); // TODO improve the way to do that
+        leftPanelEl.querySelector('.oe_snippet_clone').classList.add('d-none'); // TODO improve the way to do that
+        return this._super.apply(this, arguments);
+    },
+});
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/options.js]: Why duplicate this ?

File: addons/website/views/snippets/s_table_of_content.xml [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+                                    <p>
+                                        Create your page from scratch by dragging and dropping pre-made,
+                                        fully customizable building blocks.
+                                    </p>
+                                </div>
+                                <div class="col-lg-12 pb16">
+                                    <h1 class="text-secondary" data-anchor="true">Design features</h1>
+                                    <h4>Bootstrap based templates</h4>
+                                    <p>
+                                        Easily design your own Odoo templates thanks to clean HTML
```
Comment (qsm-odoo) [inline addons/website/views/snippets/s_table_of_content.xml]: Did not we agree on two section with no column ?

File: addons/website/views/snippets/snippets.xml [hunk 18]
```diff
@@ -51,6 +51,10 @@
                     <t t-snippet="website.s_references" t-thumbnail="/website/static/src/img/snippets_thumbs/s_references.png"/>
                     <t t-snippet="website.s_faq_collapse" t-thumbnail="/website/static/src/img/snippets_thumbs/s_faq_collapse.png"/>
                     <t t-snippet="website.s_features_grid" t-thumbnail="/website/static/src/img/snippets_thumbs/s_features_grid.png"/>
+                    <t t-snippet="website.s_popup" t-thumbnail="/website/static/src/img/snippets_thumbs/s_popup.png"/>
+                    <t t-snippet="website.s_tabs" t-thumbnail="/website/static/src/img/snippets_thumbs/s_tabs.png"/>
+                    <t t-snippet="website.s_table_of_content" t-thumbnail="/website/static/src/img/snippets_thumbs/s_table_of_content.png"/>
+                    <t t-snippet="website.s_countdown" t-thumbnail="/website/static/src/img/snippets_thumbs/s_countdown.png"/>
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml]: Rebase failed....

File: addons/website/views/snippets/snippets.xml [hunk 19]
```diff
@@ -314,7 +318,7 @@
 
     <div data-js="sizing_y"
         data-selector="section, .row > div, .parallax, .s_hr, .carousel-item, .s_rating"
-        data-exclude="section:has(> .carousel), .o_gallery .carousel-item, .s_col_no_resize.row > div"/>
+        data-exclude="section:has(> .carousel), .o_gallery .carousel-item, .s_col_no_resize.row > div, .s_table_of_content_navbar_wrap"/>
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml]: Why has the s_table_of_content_navbar_wrap element the s_col_no_resize class ? Can't we allow resizing the main columns ?

File: addons/website/static/src/snippets/s_table_of_content/000.js (hunk 11 above)
Comment (bvr-odoo) [reply addons/website/static/src/snippets/s_table_of_content/000.js]: The problem is with the fixed header. It can appear or dissapear wiith the scroll top value. So if we define the top position of the horizontal navbar with the height of the header, it can be an empty space when header is not there during the scroll

File: addons/website/static/src/snippets/s_table_of_content/000.scss (hunk 12 above)
Comment (bvr-odoo) [reply addons/website/static/src/snippets/s_table_of_content/000.scss]: because the top position of the sticky navbar is defined in js..
so the only solution to avoid this important would be to use a custom method instead of data-select-class to change the top property when it's not sticky

File: addons/website/views/snippets/s_table_of_content.xml (hunk 17 above)
Comment (bvr-odoo) [reply addons/website/views/snippets/s_table_of_content.xml]: It was not clear (SBU spec, FP spec, etc.). But I agree that it is cleaner without columns so I remove them

File: addons/website/views/snippets/snippets.xml (hunk 19 above)
Comment (bvr-odoo) [reply addons/website/views/snippets/snippets.xml]: As discussed, I changed it to allow resizing

File: addons/website/static/src/snippets/s_table_of_content/000.scss (hunk 12 above)
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_table_of_content/000.scss]: No, it's ok :+1:

File: addons/website/views/snippets/s_table_of_content.xml [hunk 20]
```diff
@@ -0,0 +1,79 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+<template id="s_table_of_content" name="Table of Content">
+    <section class="s_table_of_content pt24 pb24">
+        <div class="container table_of_content">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/s_table_of_content.xml]: That class seems unused ? remove it
